README.md
# Python Image Merger

A Python image merging script.

## Dependencies

This script uses [Pillow](https://pillow.readthedocs.io/en/stable/index.html#) for image manipulation. Please [install Pillow](https://pillow.readthedocs.io/en/stable/installation.html) before using this script.

## Usage

```python
python ImageMerger.py "--orientation=v" "--size=2" "--mode=RGB" "--fill-color=rgb(255,255,255)" "--optimize=t" "--quality=100" --verbose "--merge-path=test_files/output/test_2_vertical.png" "test_files/128/orange.png" "test_files/256/scarlet_red.png"
```

- `--orientation=[str]`: Include if you want the merge to happen horizontally (`h`) or vertically (`v`). You can place any text here, but it will only check the first letter to mach either `h` or `v`, ignoring capitalization. If the first letter doesn't match either option, it will fall back to the default. Default is `h`.
- `--size=[int]`: The number of images to fit along the width/height of the final image before changing row/column. Default is `0` (never changes the row/column).
- `--mode=[str]`: A string defining the type and depth of the pixels in the final image. Refer to [Pillow modes](https://pillow.readthedocs.io/en/stable/handbook/concepts.html#concept-modes) for all possible options. Default is `RGBA`.
- `fill-color=[str]`: A string defining the final image's background color (only visible when images have mismatched sizes, or not enough images to fill the final row/column). Refer to [Pillow color names](https://pillow.readthedocs.io/en/stable/reference/ImageColor.html#color-names) for all possible options. Default is full transparency (`rgba(0, 0, 0, 0)` / `#00000000`).
- `--optimize=[str]`: A string defining if you want the final image to have [optimized encoder settings](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html#png) (`t` / `y`) or not (`f` / `n`). You can place any text here, but it will only check the first letter to mach either `t`, `y`, `f`, or `n`, ignoring capitalization. If the first letter doesn't match either option, it will fall back to the default. Default is to optimize (`t` / `y`).
- `--quality=[int]`: An int defining the level of quality of the final image, ranging from `0` (worst) to `100` (best). Will only work on certain types of images, such as JPEG and TIFF. Default is `75`.
- `--verbose`: Include if you want the program to output to the console the value of all setup variables.
- `--merge-path=[str]`: The path where to store the final merged image, relative to the current running directory. If no value is provided, it will output the merge to `merge.png` at the current working directory.

Every image path (relative to the current running directory) listed after these arguments will add the corresponding image in that order to the final image.

## Output Examples

![2 Images (Default)](test_files/output/test_2_default.png)

![2 Images (Horizontal)](test_files/output/test_2_horizontal.png)

![2 Images (Vertical)](test_files/output/test_2_vertical.png)

![3 Images (Horizontal)](test_files/output/test_3_horizontal.png)

![4 Images (Vertical 2)](test_files/output/test_4_vertical_2.png)

![9 Images (Horizontal 3)](test_files/output/test_9_horizontal_3.png)


CHANGELOG.md
# Changelog

## [2.0.1] - 2022-09-10

### Added

### Changed

- Changed the symbols outputed on the sucess and failure console messages.
- Updated README.md.

### Removed

## [2.0.0] - 2022-09-09

### Added

- `--size` to specify rows/columns to be able to make more kinds of merges instead of just strips.
- `--quality` to specify the quality of the image in encodings such as JPEG and TIFF.
- Smaller test images to test merges with different image sizes.
- Output folder with merge samples.

### Changed

- `--vertical` is now `--orientation`, which accepts either `h` or `v`.
- Updated README.md.

### Removed

- CSV file support was dropped.

## [1.0.0] - 2022-09-09

### Added

- Initial code.
- Test images.


ImageMerger.py
import csv
import getopt
import os
import sys

from PIL import Image, ImageColor


def merge_images(orientation: str, size: int, mode: str, fill_color: tuple, image_paths: list[str]) -> tuple[Image.Image, bool]:
    try:
        images = []

        final_width = 0
        final_height = 0

        current_width = 0
        current_height = 0

        current_size = 0

        for image_path in image_paths:
            image = Image.open(image_path)
            images.append(image)

            if size == 0 or current_size < size:
                current_size += 1
            else:
                final_width = max(final_width, current_width)
                final_height = max(final_height, current_height)

                current_width = 0
                current_height = 0

                current_size = 1

            if orientation == "v":
                current_width = max(current_width, final_width + image.width)
                current_height += image.height
            else:
                current_width += image.width
                current_height = max(
                    current_height, final_height + image.height)

        final_width = max(final_width, current_width)
        final_height = max(final_height, current_height)

        mergedImage = Image.new(mode=mode, size=(
            final_width, final_height), color=fill_color)

        current_size = 0

        current_primary_pixel = 0
        current_secondary_pixel = 0
        next_secondary_pixel = 0

        for image in images:
            if size == 0 or current_size < size:
                current_size += 1
            else:
                current_size = 1

                current_primary_pixel = 0
                current_secondary_pixel = next_secondary_pixel
                next_secondary_pixel = 0

            if orientation == "v":
                mergedImage.paste(
                    image, (current_secondary_pixel, current_primary_pixel))

                current_primary_pixel += image.height
                next_secondary_pixel = max(
                    next_secondary_pixel, current_secondary_pixel + image.width)
            else:
                mergedImage.paste(
                    image, (current_primary_pixel, current_secondary_pixel))

                current_primary_pixel += image.width
                next_secondary_pixel = max(
                    next_secondary_pixel, current_secondary_pixel + image.height)

        return mergedImage, True
    except:
        return None, False


def merge_and_save(orientation: str, size: int, mode: str, fill_color: tuple, optimize: bool, quality: int, merge_path: str, image_paths: list[str]) -> tuple[Image.Image, bool]:
    try:
        finalImage, success = merge_images(
            orientation=orientation, size=size, mode=mode, fill_color=fill_color, image_paths=image_paths)
        if success:
            dirname = os.path.dirname(merge_path)
            if len(dirname) > 0:
                os.makedirs(dirname, exist_ok=True)
            finalImage.save(optimize=optimize, quality=quality, fp=merge_path)
            return finalImage, True
        else:
            return None, False
    except:
        return None, False


def main(argv):
    orientation = "h"
    size = 0
    mode = "RGBA"
    fill_color = "#00000000"
    optimize = True
    quality = 75

    verbose = False

    merge_path = "merge.png"

    option_values, image_paths = getopt.getopt(args=argv, shortopts="", longopts=[
                                               "orientation=", "size=", "mode=", "fill-color=", "optimize=", "quality=", "verbose", "merge-path="])

    for option_name, option_value in option_values:
        if option_name == "--orientation":
            try:
                if option_value[0].lower() == "v":
                    orientation = "v"
            except:
                pass

        elif option_name == "--size":
            try:
                size = int(option_value)
            except:
                pass

        elif option_name == "--mode":
            mode = option_value

        elif option_name == "--fill-color":
            fill_color = ImageColor.getrgb(option_value)

        elif option_name == "--optimize":
            try:
                if option_value[0].lower() in ["f", "n"]:
                    optimize = False
            except:
                pass

        elif option_name == "--quality":
            try:
                quality = int(option_value)
            except:
                pass

        elif option_name == "--verbose":
            verbose = True

        elif option_name == "--merge-path":
            merge_path = option_value

    if verbose:
        print("--- Merge Arguments ---")
        print("orientation:", orientation)
        print("       size:", size)
        print("       mode:", mode)
        print(" fill_color:", fill_color)
        print("   optimize:", optimize)
        print("    quality:", quality)
        print(" merge_path:", merge_path)
        print("image_paths:", image_paths)
        print("-----------------------")

    _, success = merge_and_save(orientation=orientation, size=size, mode=mode, fill_color=fill_color,
                                optimize=optimize, quality=quality, merge_path=merge_path, image_paths=image_paths)

    print("Merge", image_paths, "into", merge_path,
          "SUCCESS ☑" if success else "FAILURE ☒")


if __name__ == "__main__":
    main(sys.argv[1:])
